### PR TITLE
Improve the type analysis

### PR DESCRIPTION
**Pull request by bjorng**
Refactor the type analysis modules in the Erlang compiler to
be able to derive type information from relational operators
in guards.

Consider this function:

    f(A) when is_integer(A), 0 =< A, A =< 1000 ->
        A + 1.

From the guard, a human can easily figure out that `A` must be
an integer in the range 0 through 1000. The compiler in OTP 25
could not:

    {test,is_integer,{f,1},[{x,0}]}.
    {test,is_ge,{f,1},[{tr,{x,0},{t_integer,any}},{integer,0}]}.
    {test,is_ge,{f,1},[{integer,1000},{tr,{x,0},{t_integer,any}}]}.
    {gc_bif,'+',{f,0},1,[{tr,{x,0},{t_integer,any}},{integer,1}],{x,0}}.

With this commit, the compiler generates the following code:

    {test,is_integer,{f,1},[{x,0}]}.
    {test,is_ge,{f,1},[{tr,{x,0},{t_integer,any}},{integer,0}]}.
    {test,is_ge,{f,1},[{integer,1000},{tr,{x,0},{t_integer,{0,'+inf'}}}]}.
    {gc_bif,'+',{f,0},1,[{tr,{x,0},{t_integer,{0,1000}}},{integer,1}],{x,0}}.

The compiler can now also derive better types for the following function:

    g(A) when 0 =< A, A =< 1000 ->
         A + 1.

Since `A` is sandwiched between two numbers, `A` must be a number:

    {test,is_ge,{f,3},[{x,0},{integer,0}]}.
    {test,is_ge,
          {f,3},
          [{integer,1000},
           {tr,{x,0},
               {t_union,{t_atom,any},
                        {t_list,any,any},
                        {t_number,{0,'+inf'}},
                        {t_tuple,0,false,#{}},
                        other}}]}.
    {gc_bif,'+',{f,0},1,[{tr,{x,0},{t_number,{0,1000}}},{integer,1}],{x,0}}.

The JIT will generate slightly better code for known numeric operands than
for unknown operands.

As part of this commit, the following major changes were made:

* Ranges for integers can now extend to infinity at one endpoint. That
  is, a range can go from negative infinity to some integer, or from
  some integer to positive integer. That change allows the compiler to
  represent the type for the variable `A` when `A` is known to be an
  integer and `A >= 0` is true.

* Changed the number type to also include a (potentially infinite) range.

* Introduced an `other` type as a mean to construct a union type that
  is equivalent to the `any` type. That makes it possible to represent
  the type for `A` when all we know is that `A >= 0` (see the type in
  in the second `is_ge` test in the generated code for `g/1` above).

* The API for the `beam_bounds` module was changed to avoid having to
  create functions with the same name as operators and BIFs. Having
  functions named `min` and `max` in the module would have been
  error-prone and painful.

TO DO:

* More tests to improve coverage of some the new code.

* Investigate whether implementing subtracting of numeric types
  could lead to better types and more optimizations.

### CODE REVIEW

File: lib/compiler/src/beam_call_types.erl [hunk 1]
```diff
@@ -40,20 +40,43 @@
...
-  when S >= 0 ->
-    yes;
-will_succeed(erlang, 'bsl', [#t_integer{}=LHS,#t_integer{elements={S,_}}])
-  when S < 64 ->
-    succeeds_if_smallish(LHS);
+will_succeed(erlang, Op, [LHS, RHS]) when Op =:= 'div';
+                                          Op =:= 'rem' ->
+    case {meet(LHS, #t_integer{}), meet(RHS, #t_integer{})} of
+        {#t_integer{elements={_,_}=LHS},
+         #t_integer{elements={Div,_}=RHS}}
```
Comment (jhogberg) [inline lib/compiler/src/beam_call_types.erl]: Sorry, I bungled this part :sweat_smile: 

```suggestion
        {#t_integer{elements={_,_}}=LHS,
         #t_integer{elements={Div,_}}=RHS}
```

File: lib/compiler/src/beam_call_types.erl [hunk 2]
```diff
@@ -40,20 +40,43 @@
...
+            'yes';
+        {#t_integer{}, #t_integer{}} ->
+            'maybe';
+        {_, _} ->
+            no
+    end;
+will_succeed(erlang, 'bsr', [LHS, RHS]) ->
+     case {meet(LHS, #t_integer{}), meet(RHS, #t_integer{})} of
+         {#t_integer{elements={_,_}=LHS},
+          #t_integer{elements={Shift,_}}=RHS}
```
Comment (jhogberg) [inline lib/compiler/src/beam_call_types.erl]:
```suggestion
         {#t_integer{elements={_,_}}=LHS,
          #t_integer{elements={Shift,_}}=RHS}
```

File: lib/compiler/src/beam_call_types.erl [hunk 3]
```diff
@@ -877,50 +947,40 @@ types(_, _, Args) ->
...
 arith_type(_Op, _Args) ->
     any.
 
 %%
 %% Function-specific helpers.
 %%
 
-mixed_arith_types([FirstType | _]=Args0) ->
+mixed_arith_types(Args0) ->
+    [FirstType|_] = Args = [beam_types:normalize(A) || A <- Args0],
```
Comment (jhogberg) [inline lib/compiler/src/beam_call_types.erl]:
```suggestion
    [FirstType|_] = Args = [normalize(A) || A <- Args0],
```

File: lib/compiler/src/beam_types.erl [hunk 4]
```diff
@@ -475,7 +483,7 @@ is_boolean_type(_) ->
 
 -spec is_numerical_type(type()) -> boolean().
 is_numerical_type(#t_integer{}) -> true;
-is_numerical_type(number) -> true;
+is_numerical_type(#t_number{}) -> true;
```
Comment (jhogberg) [inline lib/compiler/src/beam_types.erl:486]: Should we treat `#t_float{}` as numerical now too?

File: lib/compiler/src/beam_types.erl [hunk 5]
```diff
@@ -973,8 +989,27 @@ lub(#t_tuple{size=SzA,elements=EsA}, #t_tuple{size=SzB,elements=EsB}) ->
...
-lub(_T1, _T2) ->
-    %%io:format("~p ~p\n", [_T1,_T2]),
+lub(T1, T2) ->
+    %%io:format("~p ~p\n", [T1,T2]),
+    case is_other(T1) andalso is_other(T2) of
+        true -> other;
+        false -> any
+    end.
+
+is_other(other) -> true;
```
Comment (jhogberg) [inline lib/compiler/src/beam_types.erl]: I think we need to add this to `beam_types_SUITE`, ensuring that `is_other(T) =:= true` can never meet any other types and vice versa.

An alternative is to express this in terms of `meet/2`, meeting the most general union type where `other=none`. If the result of that is `none`, we know that it is `other`.

File: lib/compiler/src/beam_types.erl [hunk 6]
```diff
@@ -1011,6 +1046,67 @@ gcd(A, B) ->
...
+    #t_integer{};
+integer_from_range({'-inf',Max}) ->
+    #t_integer{elements={'-inf',ceil(Max)}};
+integer_from_range({Min,'+inf'}) ->
+    #t_integer{elements={floor(Min),'+inf'}};
+integer_from_range({Min,Max}) ->
+    #t_integer{elements={floor(Min),ceil(Max)}}.
+
+number_from_range(none) ->
+    none;
```
Comment (jhogberg) [inline lib/compiler/src/beam_types.erl]: Isn't this handled by `integer_from_range/1` too?

File: lib/compiler/src/beam_types.erl (hunk 4 above)
Comment (bjorng) [reply lib/compiler/src/beam_types.erl:486]: Perhaps, but we would have to check all callers to that it is OK. I will not do that now.

File: lib/compiler/src/beam_types.erl (hunk 6 above)
Comment (bjorng) [reply lib/compiler/src/beam_types.erl]: Yes, I intended to remove it but forgot.

File: lib/compiler/src/beam_types.erl (hunk 5 above)
Comment (bjorng) [reply lib/compiler/src/beam_types.erl]: I went for your alternative approach. Does that mean that the existing test case will find bug in the handling of `other` or should I still try to write some new test case?

Comment (jhogberg) [reply lib/compiler/src/beam_types.erl]: Yeah, I think the existing model will find related errors.

**Pull request closed without merging**